src/coreclr/src/jit/lower.cpp

```diff
@@ -3024,6 +3056,132 @@ void Lowering::LowerRet(GenTreeUnOp* ret)
         BlockRange().InsertBefore(ret, bitcast);
         ContainCheckBitCast(bitcast);
     }
+#if !FEATURE_MULTIREG_RET
+    else
+    {
+#ifdef DEBUG
+        if (ret->TypeGet() != TYP_VOID)
+        {
+            if (varTypeIsStruct(ret->TypeGet()) != varTypeIsStruct(ret->gtGetOp1()->TypeGet()))
+            {
+
+                GenTree* retVal = ret->gtGetOp1();
+                if (retVal->OperIs(GT_LCL_VAR))
+                {
+                    // This could happen if we lied about op1 type during struct promotion,
```

Comment:
I wouldn't describe this as lying about the type, but rather as returning the field instead of the full struct.
I'm not sure these lengthy comments add a lot of value. I think it would be better just to assert that we have compatible types.